lib/chat/chat_messenger.dart

```diff
@@ -37,24 +37,87 @@ class ChatMessengerState extends State<ChatMessenger> {
 
   Store<AppState> store;
 
+  List<DocumentSnapshot> _messagesSnapshots;
+  bool _isLoading = false;
+  Message lastMessage;
   String recipient;
 
   File imageFile;
   String imageUrl;
 
   final TextEditingController textInputController = TextEditingController();
+  final ScrollController scrollController = ScrollController();
   final ImagePicker imagePicker = ImagePicker();
 
+  @override
+  void setState(fn) {
+    if (mounted) {
+      super.setState(fn);
+    }
+  }
+
   @override
   void initState() {
     super.initState();
     imageUrl = '';
+
+    scrollController.addListener(() {
+      if (scrollController.position.atEdge) {
+        if (scrollController.position.pixels != 0) {
+          var message = Message.fromJson(
+            _messagesSnapshots[_messagesSnapshots.length - 1]
+          );
+          Firestore.instance
+            .collection('messages')
+            .document(widget.groupChatId)
+            .collection(widget.groupChatId)
+            .where('timestamp', isLessThan: message.timestamp)
+            .orderBy('timestamp', descending: true)
+            .limit(20)
+            .getDocuments()
+            .then((snapshot) {
+              setState(() {
+                loadNewMessages();
+                _messagesSnapshots.addAll(snapshot.documents);
+              });
+            });
```

Comment:
You might be able to add this in `message_service.dart` if you do something like:

In `message_service.dart`:
```
   Future<QuerySnapshot> functionName(String groupChatId, String timestamp) {
       return  Firestore.instance
            .collection('messages')
            .document(groupChatId)
            .collection(groupChatId)
            .where('timestamp', isLessThan: timestamp)
            .orderBy('timestamp', descending: true)
            .limit(20)
            .getDocuments()
   }
```

In `chat_messenger.dart`:
```
    functionName(widget.groupChatId, message.timestamp)
        .then((snapshot) {
              setState(() {
                loadNewMessages();
                _messagesSnapshots.addAll(snapshot.documents);
              });
         });
```